### PR TITLE
Make conditionals work in debug mode

### PR DESCRIPTION
**Pull request by klecki**
## Description:
Enable using conditionals and Split/Merge in Debug Mode.
Keep batch size tracking when feasible, skip some checks when the tracking is not possible.
TODO:
 - [x] Add more tests - possibly convert all conditionals tests into debug mode.

### Affected modules and functionalities:
Debug Mode, Conditionals

### Tests:

- [ ] Existing tests apply
- [x] New tests added
  - [x] Python tests
 <more to be added>
  - [ ] GTests
  - [ ] Benchmark
  - [ ] Other
- [ ] N/A

### Documentation
- [ ] Existing documentation applies
- [ ] Documentation updated
  - [ ] Docstring
  - [ ] Doxygen
  - [ ] RST
  - [ ] Jupyter
  - [ ] Other
- [ ] N/A

#### Requirements
- [ ] Implements new requirements
- [ ] Affects existing requirements
- [ ] N/A

**REQ IDs**: N/A

**JIRA TASK**: N/A

### CODE REVIEW

**Comment by klecki** [comment]
!build

File: dali/python/nvidia/dali/_debug_mode.py [hunk 1]
```diff
@@ -23,12 +23,13 @@
...
+from nvidia.dali import _conditionals
 from nvidia.dali._utils.eager_utils import _Classification, _transform_data_to_tensorlist
 from nvidia.dali.data_node import DataNode as _DataNode, _check
 from nvidia.dali.fn import _to_snake_case
-from nvidia.dali._utils.external_source_impl import (
-    get_callback_from_source as _get_callback_from_source,
-    accepted_arg_count as _accepted_arg_count)
+from nvidia.dali._utils.external_source_impl import (get_callback_from_source as
+                                                     _get_callback_from_source, accepted_arg_count
+                                                     as _accepted_arg_count)
```
Comment (mzient) [inline dali/python/nvidia/dali/_debug_mode.py:32]: Sorry, but this is really eye-hostile.
```suggestion
                                                     _get_callback_from_source,
                                                     accepted_arg_count as _accepted_arg_count)
```

Comment (mzient) [inline dali/python/nvidia/dali/_debug_mode.py:32]: Can you make it more human friendly? (like: put X as Y in one line?)
```suggestion
from nvidia.dali._utils.external_source_impl import (
    get_callback_from_source as _get_callback_from_source,
    accepted_arg_count as _accepted_arg_count)
```

File: dali/python/nvidia/dali/_debug_mode.py [hunk 2]
```diff
@@ -46,6 +47,13 @@ def __str__(self):
     __repr__ = __str__
 
     def gpu(self):
+        if _conditionals.conditionals_enabled():
+            # Treat it the same way as regular operator would behave
+            [self_split], _ = _conditionals.apply_conditional_split_to_args([self], {})
+            transferred_node = DataNodeDebug(self._data._as_gpu(), self_split.name, "gpu",
```
Comment (mzient) [inline dali/python/nvidia/dali/_debug_mode.py:53]: What does "transferred" mean here? Or should it be "transformed"?

File: dali/python/nvidia/dali/_debug_mode.py [hunk 3]
```diff
@@ -46,6 +47,13 @@ def __str__(self):
...
+                                             self_split.source)
+            _conditionals.register_data_nodes(transferred_node, [self])
```
Comment (mzient) [inline dali/python/nvidia/dali/_debug_mode.py:55]: Nitpick: is it really always transferred? If the node is already on GPU, it's a no-op, so perhaps calling it a `gpu_node` would be more accurate.

File: dali/python/nvidia/dali/_debug_mode.py (hunk 1 above)
Comment (klecki) [reply dali/python/nvidia/dali/_debug_mode.py:32]: For me the more human friendly formatting is the one I don't need to do.

File: dali/python/nvidia/dali/_debug_mode.py (hunk 3 above)
Comment (klecki) [reply dali/python/nvidia/dali/_debug_mode.py:55]: I see a typo as well, I should copy the split node, adjusted + a shortcut.

File: dali/python/nvidia/dali/types.py [hunk 4]
```diff
@@ -518,22 +518,9 @@ def _convert(x, type):
...
-                      fdata=fdata,
-                      idata=idata,
-                      shape=shape,
-                      dtype=dtype,
-                      layout=layout,
-                      **constructor_args)
-    return op(**call_args)
+    op = fn.constant(device=device, fdata=fdata, idata=idata, shape=shape, dtype=dtype,
+                     layout=layout, **kwargs)
+    return op
```
Comment (mzient) [inline dali/python/nvidia/dali/types.py:523]: :+1: 
You can take it one step further:
```suggestion
    return fn.constant(device=device, fdata=fdata, idata=idata, shape=shape, dtype=dtype,
                       layout=layout, **kwargs)
```

File: dali/test/python/conditionals/test_pipeline_conditionals.py [hunk 5]
```diff
@@ -581,13 +591,17 @@ def scoped_transform_pipe():
...
+
+
+def impl_test_arg_inputs_scoped_uninitialized(additional_additional_kwargs={}):
     test_data_root = get_dali_extra_path()
     caffe_db_folder = os.path.join(test_data_root, 'db', 'lmdb')
     bs = 10
     kwargs = {"batch_size": bs, "num_threads": 4, "device_id": 0}
 
-    @experimental.pipeline_def(enable_conditionals=True, **kwargs)
+    @experimental.pipeline_def(enable_conditionals=True, **kwargs, **additional_additional_kwargs)
```
Comment (banasraf) [inline dali/test/python/conditionals/test_pipeline_conditionals.py:604]: why `additional_additional_kwargs`?

File: dali/python/nvidia/dali/_debug_mode.py (hunk 1 above)
Comment (klecki) [reply dali/python/nvidia/dali/_debug_mode.py]: (Done btw)

File: dali/test/python/test_pipeline_debug.py [hunk 6]
```diff
@@ -649,3 +657,128 @@ def test_nan_check():
...
+        pred = fn.random.coin_flip(seed=42, dtype=types.BOOL)
+        rng1 = fn.random.coin_flip(seed=1)
+        if pred:
+            output = fn.copy(rng1.get())
+        else:
+            output = rng1 + 10
+        return pred, output
+
+    with assert_raises(
+            ValueError, glob=("Debug mode for conditionals doesn't allow for modification of"
```
Comment (mzient) [inline dali/test/python/test_pipeline_debug.py:778]:
```suggestion
            ValueError, glob=("Debug mode for conditional statements doesn't allow for modification of"
```
"conditionals" is our internal jargon.

File: dali/python/nvidia/dali/_debug_mode.py [hunk 7]
```diff
@@ -322,6 +363,11 @@ def __init__(self, op_class, op_name, pipe, source_context, next_logical_id, bat
                     raise ValueError("All argument lists for Multiple Input Sets used "
                                      f"with operator '{op_name}' must have the same length.")
             self._inputs_classification.append(classification)
+
+        if _conditionals.conditionals_enabled():
+            if input_set_len != -1:
+                raise ValueError("Multiple input sets are not supported with conditionals.")
```
Comment (mzient) [inline dali/python/nvidia/dali/_debug_mode.py:369]:
```suggestion
                raise ValueError("Multiple input sets are not supported in conditional statements.")
```

File: dali/python/nvidia/dali/_debug_mode.py [hunk 8]
```diff
@@ -531,10 +624,30 @@ def run(self, inputs, kwargs):
...
+                                               classification.data, 'Argument', key)
                 call_args[key] = classification.data
 
+        if _conditionals.conditionals_enabled():
+            # Did we manage to succefuly extract a input batch, but the input was not directly
+            # produced by DALI
+            # TODO(klecki): Add better handling of constant nodes for conditionals in debug mode.
+            for i, classification in enumerate(self._inputs_classification):
+                if classification.is_batch and not classification.was_data_node:
+                    raise ValueError(f"Debug mode for conditionals doesn't allow for modification"
```
Comment (mzient) [inline dali/python/nvidia/dali/_debug_mode.py:637]:
```suggestion
                    raise ValueError(f"Debug mode for conditional statements doesn't allow for modification"
```

File: dali/python/nvidia/dali/_debug_mode.py [hunk 9]
```diff
@@ -531,10 +624,30 @@ def run(self, inputs, kwargs):
...
+                                     f" of operator outputs by libraries other than DALI or"
+                                     f" tracking the TensorLists extracted via `.get()`. Expected"
+                                     f" `DataNodeDebug` as an input, got"
+                                     f" {type(classification.original)} at input {i}.")
+
+            for key, classification in self._kwargs_classification.items():
+                if classification.is_batch and not classification.was_data_node:
+                    raise ValueError(f"Debug mode for conditionals doesn't allow for modification"
```
Comment (mzient) [inline dali/python/nvidia/dali/_debug_mode.py:645]:
```suggestion
                    raise ValueError(f"Debug mode for conditional statements doesn't allow for modification"
```

**Review by mzient — APPROVED** [review]
Please change the messages not to use our jargon - otherwise good.

File: dali/python/nvidia/dali/_debug_mode.py (hunk 8 above)
Comment (klecki) [reply dali/python/nvidia/dali/_debug_mode.py:637]: done

File: dali/test/python/test_pipeline_debug.py (hunk 6 above)
Comment (klecki) [reply dali/test/python/test_pipeline_debug.py:778]: done

File: dali/python/nvidia/dali/_debug_mode.py (hunk 7 above)
Comment (klecki) [reply dali/python/nvidia/dali/_debug_mode.py:369]: done

**Comment by klecki** [comment]
!build

**Pull request merged**